### DISCUSSION TITLE
Contribution workflow: new keyboard, new layout

### DISCUSSION THREAD

**Discussion by lmlask** [DreymaR/BigBagKbdTrixPKL#33] [Q&A] [answered]
Hi, DreymaR! I've been a Workman user for some time and love your Extend layer. I believe the Big Bag is the single best implementation of PKL today, but since it doesn't feature Workman I had to program my own utilities (for several OS's, mind you, as I use 3 Linux distros at home and Windows at work). I'd like to contribute and implement Workman and variations for the Big Bag, but I'm failing to see how some things work. I can just copy and paste another layout folder and change everything, but EPKL still doesn't recognize the layout variations, mods etc of I just do that. Is the program looking for a table somewhere which I can edit and fill with the proper codes?

One more contribution I'd like to make: the ABNT keyboard. While ANS has two big shift keys and ISO has the left shift shrinked to fit the ISO key, ABNT does the same with the right shift to get an extra key there (the Enter key is the small L just like the ISO keyb). I can get key codes and figure out permutations so the program will recognize that extra key (which is AB11 in Linux/XKB terms).

I guess what I'm asking is, can you provide a general contributing guide, not for git stuff but for which files I should be changing in what order? As you implemented many layouts I'm sure you have a pretty good workflow which it'd be nice to have too.

Thank you and have a nice day!

**Comment by DreymaR** [comment, marked as answer]
Thanks for the praise!

The ABNT keys are covered in the EPKL_Layouts_Default file. Since not everyone has them, that's where they went. You can actually map things to them in the KeyMapper GUI, but you'll have to write in their names manually then.

As for Workman... Sorry, but I'm fundamentally against supporting that layout. It has design flaws that put me off it, together with the arrogance and cluelessness of its maker, and I honestly think people should use a layout like Colemak-DH instead for the far better design with lower same-finger bigram stats. I think it's a shame it's been so aggressively promoted and newcomers see it as "Colemak done better" until they hopefully eventually learn their mistake.

I've added several oddball and even joke layouts to EPKL so this may strike you as inconsistent. But some layouts that I view as flawed such as Minimak, Norman and Workman must make do without support from me I'm afraid, as I don't want to draw any further attention to them in an already too crowded layout landscape. Again, sorry.

However, you are of course free to make such a layout folder for yourself and distribute it as you see fit! Now, there is no guide like the one you wish for I'm afraid. I'm hoping for someone with good explaining skills to make such guides, as there used to be one but it's now mostly outdated. I think you may be on the right track, but you have to be careful with your file and folder naming. Let's say you make a Workman main layout folder, fine.  You'll have to use 'Wor' as its 3-letter abbreviation then, even if Colemak is abbreviated to 'Cmk' as you can see. You could add your layout in the Tables file to use another 3LA, but it wouldn't be worth it.

Basically, the answer is that naming rules are important, and mostly it isn't about table entries for layout variants/mods. I think you'll get the hang of things by studying the examples of other layouts. Note that some layouts like Colemak have subfolders to organize variants, but that isn't necessary and other layouts with fewer variants don't have subfolders.

**Reply by lmlask** [reply]
Thank you for the reply. Of course, I understand your choice not to support it, although I personally am for giving people options. While I'm a Workman dude, I also agree Colemak should probably be the standard for replacing QWERTY.

I'll set up a fork for personal use then, if you ever change your mind about it (today I would probably have learned Colemak, but I don't think I want to learn a layout again lol) it'll be there.

Thanks again and happy coding!

**Reply by DreymaR** [reply]
Thanks for your understanding! It's in part about time and effort, every time I want to change something there's now hundreds of files to change and complexity just keeps growing. And I'm just one man on this project.

Hope you get it working! You can ask if you're stuck somewhere. But I can't take the time and effort to write up a proper guide now.
